src/v/archival/archival_metadata_stm.cc

```diff
@@ -1018,81 +946,94 @@ ss::future<> archival_metadata_stm::apply(const model::record_batch& b) {
+                    apply_update_start_offset(
+                      serde::from_iobuf<update_start_offset_cmd::value>(
+                        r.release_value()));
+                    break;
+                case cleanup_metadata_cmd::key:
+                    apply_cleanup_metadata();
+                    break;
+                case mark_clean_cmd::key:
+                    apply_mark_clean(serde::from_iobuf<mark_clean_cmd::value>(
+                      r.release_value()));
+                    break;
+                case truncate_archive_init_cmd::key:
+                    apply_truncate_archive_init(
+                      serde::from_iobuf<truncate_archive_init_cmd::value>(
+                        r.release_value()));
+                    break;
+                case truncate_archive_commit_cmd::key: {
+                    auto cmd
+                      = serde::from_iobuf<truncate_archive_commit_cmd::value>(
+                        r.release_value());
+                    apply_truncate_archive_commit(
+                      cmd.start_offset, cmd.bytes_removed);
+                } break;
+                case update_start_kafka_offset_cmd::key:
+                    apply_update_start_kafka_offset(
+                      serde::from_iobuf<update_start_kafka_offset_cmd::value>(
+                        r.release_value()));
+                    break;
+                case reset_metadata_cmd::key:
+                    apply_reset_metadata();
+                    break;
+                case spillover_cmd::key:
+                    apply_spillover(
+                      serde::from_iobuf<spillover_cmd>(r.release_value()));
+                    break;
+                case replace_manifest_cmd::key:
+                    apply_replace_manifest(r.release_value());
+                    break;
+                case process_anomalies_cmd::key:
+                    apply_process_anomalies(r.release_value());
+                    break;
+                case reset_scrubbing_metadata::key:
+                    apply_reset_scrubbing_metadata();
+                    break;
+                case update_highest_producer_id_cmd::key:
+                    apply_update_highest_producer_id(
+                      serde::from_iobuf<update_highest_producer_id_cmd::value>(
+                        r.release_value()));
+                    break;
+                default:
+                    throw std::runtime_error(fmt_with_ctx(
+                      fmt::format,
+                      "Unknown archival metadata STM command {}",
+                      static_cast<int>(key)));
+                };
+            });
+        } catch (...) {
+            vlog(
+              _logger.error,
+              "Unexpected error while applying changes to "
+              "archival_metadata_stm: {}",
+              std::current_exception());
+            on_exit.cancel();
+            maybe_notify_waiter(std::current_exception());
```

Comment:
Thinking about this a bit more I think that we don't even need to do this. When we replicating config batch we're acquiring the lock. Then, while holding the lock we're setting the promise, replicating the batch and calling `wait_no_throw` and then waiting for the promise to be set. The entire lifecycle of the promise is done under the lock so such violation is not possible. Concurrent fiber may try to replicate something when the promise is created but not yet set but it will get stuck waiting for the lock.